st2actions/st2actions/workflows/workflows.py

```diff
@@ -78,13 +94,61 @@ def process(self, message):
             raise ValueError(msg)
 
         try:
+            with self._semaphore:
+                self._active_messages += 1
             handler_function(message)
         except Exception as e:
             # If the exception is caused by DB connection error, then the following
             # error handling routine will fail as well because it will try to update
             # the database and fail the workflow execution gracefully. In this case,
             # the garbage collector will find and cancel these workflow executions.
             self.fail_workflow_execution(message, e)
+        finally:
+            with self._semaphore:
+                self._active_messages -= 1
+
+    def shutdown(self):
+        super(WorkflowExecutionHandler, self).shutdown()
+        while self._active_messages > 0:
+            concurrency.sleep(2)
+
+        coordinator = coordination.get_coordinator()
+        member_ids = []
+        with coordinator.get_lock(SHUTDOWN_ROUTINE):
+            try:
+                member_ids = list(
+                    coordinator.get_members(WORKFLOW_ENGINE.encode("utf-8")).get()
```

Comment:
Instead of doing `WORKFLOW_ENGINE.encode("utf-8")`, can we add a method in the coordination service that get_group_id with the logic @ https://github.com/StackStorm/st2/blob/master/st2common/st2common/service_setup.py#L324-L327. That way, you can call `coordinator.get_group_id(WORKFLOW_ENGINE)`. This is a safer way to get group id so to shield from any changes to group_id naming/formatting.

Reply:
[coordination.get_members()](https://github.com/openstack/tooz/blob/master/tooz/coordination.py#L551) returns an async`CoordinatorResult` which has a [get()](https://github.com/openstack/tooz/blob/master/tooz/coordination.py#L642) method that retrieves the result of the corresponding asynchronous call.